syntax = "proto3";

service DBService {
    // Users
    rpc UsersInsert (UsersInsertRequest) returns (UsersInsertResponse);
    rpc UsersDelete (UsersDeleteRequest) returns (UsersDeleteResponse);
    rpc UsersSelect (UsersSelectRequest) returns (UsersSelectResponse);
    rpc UsersUpdate (UsersUpdateRequest) returns (UsersUpdateResponse);
    // Products
    rpc ProductsInsert (ProductsInsertRequest) returns (ProductsInsertResponse);
    rpc ProductsDelete (ProductsDeleteRequest) returns (ProductsDeleteResponse);
    rpc ProductsSelect (ProductsSelectRequest) returns (ProductsSelectResponse);
    rpc ProductsUpdate (ProductsUpdateRequest) returns (ProductsUpdateResponse);
    // Orders
    rpc OrdersInsert (OrdersInsertRequest) returns (OrdersInsertResponse);
    rpc OrdersDelete (OrdersDeleteRequest) returns (OrdersDeleteResponse);
    rpc OrdersSelect (OrdersSelectRequest) returns (OrdersSelectResponse);
    rpc OrdersUpdate (OrdersUpdateRequest) returns (OrdersUpdateResponse);
}

// The Users
message User {
    int64 id = 1;
    string sid = 2;
    string username = 3;
    string password_hash = 4;
    string email = 5;
    string created_at = 6;
}
// InsertRequest
message UsersInsertRequest {
    string sid = 1;
    string username = 2;
    string email = 3;
    string password_hash = 4;
}
// InsertResponse
message UsersInsertResponse {
    string message = 1;
    User user = 2;
}
// DeleteRequest
message UsersDeleteRequest {
    int64 id = 1;
}
// DeleteResponse
message UsersDeleteResponse {
    string message = 1;
}
// SelectRequest
message UsersSelectRequest {
    int64 id = 1;
    string sid = 2;
    string username = 3;
    string email = 4;
    string password_hash = 5;
    string created_at = 6;
}
// SelectResponse
message UsersSelectResponse {
    string message = 1;
    repeated User matched_users = 2;
}
// UpdateRequest
message UsersUpdateRequest {
    int64 id = 1;
    string sid = 2;
    string username = 3;
    string password_hash = 4;
    string email = 5;
    string created_at = 6;
}
// UpdateResponse
message UsersUpdateResponse {
    string message = 1;
    User user = 2;
}

// The Products
message Product {
    int64 id = 1;
    string name = 2;
    string description = 3;
    string category = 4;
    float price = 5;
    string slogan = 6;
    int64 stock = 7;
    string created_at = 8;
}
// InsertRequest
message ProductsInsertRequest {
    string name = 1;
    string description = 2;
    string category = 3;
    float price = 4;
    string slogan = 5;
    int64 stock = 6;
}
// InsertResponse
message ProductsInsertResponse {
    string message = 1;
    Product product = 2;
}
// DeleteRequest
message ProductsDeleteRequest {
    int64 id = 1;
}
// DeleteResponse
message ProductsDeleteResponse {
    string message = 1;
}
// SelectRequest
message ProductsSelectRequest {
    int64 id = 1;
    string name = 2;
    string description = 3;
    string category = 4;
    float price = 5;
    string slogan = 6;
    int64 stock = 7;
    string created_at = 8;
}
// SelectResponse
message ProductsSelectResponse {
    string message = 1;
    repeated Product matched_products = 2;
}
// UpdateRequest
message ProductsUpdateRequest {
    int64 id = 1;
    string name = 2;
    string description = 3;
    string category = 4;
    float price = 5;
    string slogan = 6;
    int64 stock = 7;
    string created_at = 8;
}
// UpdateResponse
message ProductsUpdateResponse {
    string message = 1;
    int64 id = 2;
}

// The Orders
message Order {
    int64 id = 1;
    int64 user_id = 2;
    int64 product_id = 3;
    int64 quantity = 4;
    float total_price = 5;
    string created_at = 6;
}
// InsertRequest
message OrdersInsertRequest {
    int64 user_id = 1;
    int64 product_id = 2;
    int64 quantity = 3;
    float total_price = 4;
}
// InsertResponse
message OrdersInsertResponse {
    string message = 1;
    Order order = 2;
}
// The OrdersDeleteRequest message
message OrdersDeleteRequest {
    int64 id = 1;
}
// The OrdersDeleteResponse message
message OrdersDeleteResponse {
    string message = 1;
}
// The OrdersSelectRequest message
message OrdersSelectRequest {
    int64 id = 1;
    int64 user_id = 2;
    int64 product_id = 3;
    int64 quantity = 4;
    float total_price = 5;
    string created_at = 6;
}
// The OrdersSelectResponse message
message OrdersSelectResponse {
    string message = 1;
    repeated Order matched_orders = 2;
}
// The OrdersUpdateRequest message
message OrdersUpdateRequest {
    int64 id = 1;
    int64 user_id = 2;
    int64 product_id = 3;
    int64 quantity = 4;
    float total_price = 5;
    string created_at = 6;
}
// The OrdersUpdateResponse message
message OrdersUpdateResponse {
    string message = 1;
    Order order = 2;
}